### DISCUSSION TITLE
Per-Player Entity Visibility

### DISCUSSION THREAD

**Discussion by blake-tisbury** [Minestom/Minestom#726] [Help] [answered]
Can you hide entities/players per player or is it only controlled from the context of the server? (Eg. If Player 1 is on a gold block they then can see player 2 if they are also standing on a gold block, but no other players can see them) 
Trying to make a dynamic culling system and wondering if it’s even possible

**Comment by TheMode** [comment, marked as answer]
`Entity#updateViewableRule(Predicate<Player>)` can be used to predicate the players able to see you.
And `Entity#updateViewerRule(Predicate<Entity>)` to predicate the entities you are able to see.

You must be aware that since the server is unable to know the precise moment these predicates change result, and that checking every tick would be expensive, updates happen on rule change (methods above) and with the non-parameterized versions `Entity#updateViewableRule()`/`Entity#updateViewerRule()`

I must also mention that tweaking visibility may prevent important optimizations from taking place (e.g. viewable packets) and therefore increasing tick time. Be sure to measure the impact.

**Reply by blake-tisbury** [reply]
Thanks a ton!
